src/external.py

```diff
@@ -0,0 +1,188 @@
+# Copyright (c) 2018-2020, Michael P. Howard
+# Copyright (c) 2021-2024, Auburn University
+# Part of azplugins, released under the BSD 3-Clause License.
+
+"""External harmonic potentials."""
+
+import hoomd
+
+from hoomd.azplugins import _azplugins
+from hoomd.data.parameterdicts import TypeParameterDict
+from hoomd.data.typeparam import TypeParameter
+from hoomd.data.parameterdicts import ParameterDict
+from hoomd.md import force
+
+
+class HarmonicBarrier(force.Force):
+    r"""Applies Harmonic Potential.
+
+    This class should not be instantiated directly. Use a derived type.
+
+    Args:
+        interface (hoomd.variant.variant): position of the interface
+        geometry (str): Drying geometry ("film" or "droplet")
+
+    Args:
+        k (float): Harmonic coefficient
+
+        offset (float): per-particle-type amount to shift in the potential
+
+    """
+
+    def __init__(self, interface=None, geometry=None):
+        super().__init__()
+
+        if geometry not in ("film", "droplet"):
+            raise ValueError(f"Unrecognized geometry: {geometry}")
+
+        param_dict = ParameterDict(interface=hoomd.variant.Variant, geometry=str)
+        param_dict["interface"] = interface
+        param_dict["geometry"] = geometry
+        param_dict.update(
+            dict(
+                interface=interface,
+                geometry=geometry,
+            )
+        )
+        self._param_dict.update(param_dict)
+
+        params = TypeParameter(
+            "params",
+            "particle_types",
+            TypeParameterDict(k=float, offset=float, len_keys=1),
+        )
+        self._add_typeparam(params)
+
+    def _attach_hook(self):
+        sim = self._simulation
+
+        cpp_class = getattr(_azplugins, self._make_cpp_class_name())
+
+        if self.geometry not in ("film", "droplet"):
+            raise ValueError(f"Unrecognized geometry: {self.geometry}")
+
+        self._cpp_obj = cpp_class(sim.state._cpp_sys_def, self.interface)
+
+        super()._attach_hook()
+
+    def _make_cpp_class_name(self):
+        cpp_class_name = self.__class__.__name__
+        if isinstance(self._simulation.device, hoomd.device.GPU):
+            cpp_class_name += "GPU"
+        return cpp_class_name
+
+
+class PlanarHarmonicBarrier(HarmonicBarrier):
+    r"""Applies a purely repulsive harmonic potential to particles
+        near an interface in planar geometry..
+
+    Args:
+        interface (`hoomd.variant`) : `z` position of the interface
```

Comment:
Backticks are used for cross referencing. The type should also be variant_like.
```suggestion
        interface (`hoomd.variant.variant_like`) : *z* position of the interface
```